controllers/aso_credential_cache.go

```diff
@@ -0,0 +1,221 @@
+	d := secret.Data
+
+	if _, hasSecret := d[config.AzureClientSecret]; hasSecret {
+		return c.cache.GetOrStoreClientSecret(
+			string(d[config.AzureTenantID]),
+			string(d[config.AzureClientID]),
+			string(d[config.AzureClientSecret]),
+			&azidentity.ClientSecretCredentialOptions{
+				ClientOptions: clientOpts,
+			},
+		)
+	}
+
+	if _, hasCert := d[config.AzureClientCertificate]; hasCert {
+		return c.cache.GetOrStoreClientCert(
+			string(d[config.AzureTenantID]),
+			string(d[config.AzureClientID]),
+			d[config.AzureClientCertificate],
+			d[config.AzureClientCertificatePassword],
+			&azidentity.ClientCertificateCredentialOptions{
+				ClientOptions: clientOpts,
+			},
+		)
+	}
+
+	if authMode := d[config.AuthMode]; config.AuthModeOption(authMode) == config.PodIdentityAuthMode {
+		return c.cache.GetOrStoreManagedIdentity(
+			&azidentity.ManagedIdentityCredentialOptions{
+				ClientOptions: clientOpts,
+				ID:            azidentity.ClientID(d[config.AzureClientID]),
+			},
+		)
+	}
+
+	return c.cache.GetOrStoreWorkloadIdentity(
+		&azidentity.WorkloadIdentityCredentialOptions{
+			ClientOptions: clientOpts,
+			TenantID:      string(d[config.AzureTenantID]),
+			ClientID:      string(d[config.AzureClientID]),
+			TokenFilePath: federatedTokenFilePath(),
+		},
+	)
+}
+
+func (c *asoCredentialCache) authTokenForGlobalASOSecret(secret *corev1.Secret, clientOpts azcore.ClientOptions) (azcore.TokenCredential, error) {
+	d := secret.Data
+
+	if workloadID, _ := strconv.ParseBool(string(d[config.UseWorkloadIdentityAuth])); workloadID {
+		return c.cache.GetOrStoreWorkloadIdentity(
+			&azidentity.WorkloadIdentityCredentialOptions{
+				ClientOptions: clientOpts,
+				TenantID:      string(d[config.AzureTenantID]),
+				ClientID:      string(d[config.AzureClientID]),
+				TokenFilePath: federatedTokenFilePath(),
+			},
+		)
+	}
+
+	if _, hasSecret := d[config.AzureClientSecret]; hasSecret {
+		return c.cache.GetOrStoreClientSecret(
+			string(d[config.AzureTenantID]),
+			string(d[config.AzureClientID]),
+			string(d[config.AzureClientSecret]),
+			&azidentity.ClientSecretCredentialOptions{
+				ClientOptions: clientOpts,
+			},
+		)
+	}
+
+	if _, hasCert := d[config.AzureClientCertificate]; hasCert {
+		return c.cache.GetOrStoreClientCert(
+			string(d[config.AzureTenantID]),
+			string(d[config.AzureClientID]),
+			d[config.AzureClientCertificate],
+			d[config.AzureClientCertificatePassword],
+			&azidentity.ClientCertificateCredentialOptions{
+				ClientOptions: clientOpts,
+			},
+		)
+	}
+
+	return c.cache.GetOrStoreManagedIdentity(
+		&azidentity.ManagedIdentityCredentialOptions{
+			ClientOptions: clientOpts,
+			ID:            azidentity.ClientID(d[config.AzureClientID]),
+		},
+	)
+}
+
+func federatedTokenFilePath() string {
+	if env, ok := os.LookupEnv("AZURE_FEDERATED_TOKEN_FILE"); ok {
+		return env
+	}
+	return "/var/run/secrets/azure/tokens/azure-identity-token"
```

Comment:
We should plan on making this a constant `"/var/run/secrets/azure/tokens/azure-identity-token"`

Reply:
I have a refactor for this queued up as part of #1077.